spec/unit/capability_spec.rb

```diff
@@ -128,18 +135,138 @@ class mod {
-      manifest = <<-MANIFEST
+    ["produces", "consumes"].each do |kw|
+      it "creates an error when #{kw} references nonexistent type" do
+        manifest = <<-MANIFEST
         Test #{kw} Cap {
           host => $hostname
         }
       MANIFEST
 
-      expect {
-        compile_to_catalog(manifest)
-      }.to raise_error(Puppet::Error,
-                       /#{kw} clause references nonexistent type Test/)
+        expect {
+          compile_to_catalog(manifest)
+        }.to raise_error(Puppet::Error,
+                         /#{kw} clause references nonexistent type Test/)
+      end
+    end
+  end
+
+  describe "exporting a capability" do
+    before(:each) do
+      make_cap_type
+    end
+
+    after :each do
+      Puppet::Type.rmtype(:cap)
+    end
+
+    it "does not add produced resources that are not exported" do
+      manifest = <<-MANIFEST
+define test($hostname) {
+  notify { "hostname ${hostname}":}
+}
+
+Test produces Cap {
+  host => $hostname
+}
+
+test { one: hostname => "ahost" }
+    MANIFEST
+      catalog = compile_to_catalog(manifest)
+      expect(catalog.resource("Test[one]")).to be_instance_of(Puppet::Resource)
+      expect(catalog.resource_keys.find { |type, _| type == "Cap" }).to be_nil
+    end
+
+    it "adds produced resources that are exported" do
+      manifest = <<-MANIFEST
+define test($hostname) {
+  notify { "hostname ${hostname}":}
+}
+
+# The $hostname in the produces clause does not refer to this variable,
+# instead, it referes to the hostname property of the Test resource
+# that is producing the Cap
+$hostname = "other_host"
+
+Test produces Cap {
+  host => $hostname
+}
+
+test { one: hostname => "ahost", export => Cap[two] }
+    MANIFEST
+      catalog = compile_to_catalog(manifest)
+      expect(catalog.resource("Test[one]")).to be_instance_of(Puppet::Resource)
+
+      caps = catalog.resource_keys.select { |type, _| type == "Cap" }
+      expect(caps.size).to eq(1)
+
+      cap = catalog.resource("Cap[two]")
+      expect(cap).to be_instance_of(Puppet::Resource)
+      expect(cap["require"]).to eq("Test[one]")
+      expect(cap["host"]).to eq("ahost")
+      expect(cap.resource_type).to eq(Puppet::Type::Cap)
+      expect(cap.tags.any? { |t| t == "producer:production" }).to eq(true)
+    end
+  end
+
+  describe "consuming a capability" do
+    before(:each) do
+      make_cap_type
+    end
+
+    after :each do
+      Puppet::Type.rmtype(:cap)
+    end
+
+    def make_catalog(instance)
+      manifest = <<-MANIFEST
+define test($hostname = nohost) {
+  notify { "hostname ${hostname}":}
+}
+
+Test consumes Cap {
+  hostname => $host
+}
+    MANIFEST
+      compile_to_catalog(manifest + instance)
+    end
+
+    def mock_cap_finding
+      cap = Puppet::Resource.new("Cap", "two")
+      cap["host"] = "ahost"
+      Puppet::Resource::CapabilityFinder.expects(:find).returns(cap)
+      cap
+    end
+
+    it "does not fetch a consumed resource when consume metaparam not set" do
```

Comment:
are there two sides to this: "fetches a consumed resource when consume metaparam set" and then the test above. if so, it would be good if they are grouped (maybe there is such a test, but it is hard to manually correlate). e.g. using a context "consume metaparam controls fetching such as", it "when not set, does not fetch a consumed resource", and it "when set, fetches a consumed resource". (such grouping helps QA review test coverage).